### PR TITLE
Fix Random thread safety and control random seed to improve testability

### LINKED ISSUE #4361: Fix Random thread safety and control random seed to improve testability
# Problem
Our usage of Random is not thread safe and when writing tests, you currently need to test for all random responses.  This makes it harder to write tests and adds minimally to the test coverage.  One method in particular is to take a transcript and turn it into declarative test to make sure the same path gets executed.  This does not work if there is any variation in .lg templates or if the random selector is used.  

# Context
Where random is used:
1) Rand function in adaptive-expressions.  This uses a single static generator which is not thread safe. There is no way to set the seed.
2) .lg randomization in Evaluator and Expander.  Each of these creates a new Random object on every call.  Creating a Random object is expensive, so this is not ideal and there is no way to set the seed.
3) Random selector creates a new Random for each instance and does support setting the seed through the schema.  This is also potentially an issue with thread safety.

# Solution
1) Create middleware in the adaptive test package to respond to a "Microsoft.SetTestOptions" event by taking the payload and putting it into conversation.TestOptions.
2) Everywhere we use random numbers (adaptive expressions, .lg, random selector) check for conversationTestOptions.RandomValue and if present use that in place of calling random.  

To support recording a transcript with the random value set we also need to add this to the emulator which will allow emitting the test event while specifying the RandomValue.  https://github.com/microsoft/BotFramework-Emulator/issues/2174

With these in place a transcript will contain the Microsoft.SetTraceOptions trace that can be included in a declarative test to emit the same activity so that the test script will have the same predictable behavior.

This solution will enable us to control the random choices made so that we can ensure that a test will run consistently.  Once we have made these improvements, we will also need to add the middle ware to the default Composer runtime.

[enhancement]

### PR DESCRIPTION
**Pull request by Danieladu**
Fixes #4361

## Description
This PR aims to accept a "Microsoft.SetTestOptions" event, and apply the options to SDK.

## Specific Changes
1. Add `setTestOptions` event
2. Save TestOptions into the conversation state. Including the Random seed and random value.
3. Achieve the state from "dialogContext.State", and apply it to SDK code.
4. Adjust random generator code to use the random in "TestOptions".

## How to test
1. Start project "Microsoft.Bot.Builder.TestBot.Json"
2. Input 26 to see the dialog output.

## Progress
- [x] Adaptive-expression
- [x] LG Evaluator/Expander
- [ ] Random selector

### CODE REVIEW

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/TestOptions.cs [hunk 1]
```diff
@@ -0,0 +1,25 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Microsoft.Bot.Builder.Dialogs.Adaptive
+{
+    /// <summary>
+    /// test options.
```
Comment (chrimc62) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/TestOptions.cs:8]: >test [](start = 8, length = 4)

Initial cap.

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/EvaluationOptions.cs [hunk 2]
```diff
@@ -43,6 +43,7 @@ public EvaluationOptions()
             NullSubstitution = null;
             LineBreakStyle = null;
             Locale = null;
+            Random = null;
```
Comment (chrimc62) [inline libraries/Microsoft.Bot.Builder.LanguageGeneration/EvaluationOptions.cs:46]: This is an anti-pattern.  Either assign to each definition or have this, but both is weird.  For example you assign random in the get/set below to a random object, but here you set it to null.  If the default is null you don't need to set anything and if the default is a random object it should be that.

File: tests/Microsoft.Bot.Builder.TestBot.Json/TestOptionsMiddleware.cs [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Bot.Builder.Dialogs.Adaptive;
+using Microsoft.Bot.Schema;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Bot.Builder.TestBot.Json
+{
+    public class TestOptionsMiddleware : IMiddleware
```
Comment (chrimc62) [inline tests/Microsoft.Bot.Builder.TestBot.Json/TestOptionsMiddleware.cs:12]: >TestOptionsMiddleware [](start = 17, length = 21)

This should be part of Microsoft.Bot.Builder.Dialogs.Adaptive.Testing to be used by everyone.

File: tests/Microsoft.Bot.Builder.TestBot.Json/TestBotHttpAdapter.cs [hunk 4]
```diff
@@ -21,6 +21,8 @@ public class TestBotHttpAdapter : BotFrameworkHttpAdapter
             : base(configuration, credentialProvider)
         {
             this.Use(new RegisterClassMiddleware<IConfiguration>(configuration));
+            this.Use(new ConvertMessageToEventMiddleware());
+            this.Use(new TestOptionsMiddleware(conversationState));
```
Comment (chrimc62) [inline tests/Microsoft.Bot.Builder.TestBot.Json/TestBotHttpAdapter.cs:25]: Why is this needed?

File: tests/Microsoft.Bot.Builder.TestBot.Json/Samples/RandomGenerator/RandomGenerator.main.dialog [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+        {
+            "$kind": "Microsoft.OnBeginDialog",
+            "actions": [
+                {
+                    "$kind": "Microsoft.SendActivity",
+                    "activity": "${rand(1, 10)}"
+                }
+            ]
+        }
+    ]
```
Comment (chrimc62) [inline tests/Microsoft.Bot.Builder.TestBot.Json/Samples/RandomGenerator/RandomGenerator.main.dialog:14]: I don't understand this.

**Comment by chrimc62** [comment]
Good progress.  A few things:
1) Specifying a single random with a seed is useful, but even more useful is if you can specify the single number to return every place you want a random number.  The reason is that it makes for a much more stable test that is not dependent on how many times you call random.  Both should be in your options.
2) To set the test options you should be using a custom setTestOptions event using an emitEvent action where eventName = "setTestOptions" and eventValue={ randomSeed: <seed for random>, randomValue: <constant to return in every place you want random>}.  This is the event that middle ware should pick up and make available ideally through conversation.testOptions memory.  This is a general mechanism where over time we can add more information to the object and either bake in uses like here or even have templates output extra information if you want to do custom testing stuff.

File: libraries/AdaptiveExpressions/BuiltinFunctions/Rand.cs [hunk 6]
```diff
@@ -43,9 +41,18 @@ private static EvaluateExpressionDelegate Evaluator()
                             }
                             else
                             {
-                                lock (_randomizerLock)
+                                if (options.RandomValue != null)
+                                {
+                                    value = options.RandomValue;
```
Comment (chrimc62) [inline libraries/AdaptiveExpressions/BuiltinFunctions/Rand.cs:46]: >options.RandomValue; [](start = 44, length = 20)

This needs to be adjusted to be between min/max, i.e. min + (options.RandomValue % (max - min))

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/AdaptiveDialog.cs [hunk 7]
```diff
@@ -383,6 +383,13 @@ protected virtual async Task<bool> ProcessEventAsync(ActionContext actionContext
                         activity = ObjectPath.GetPathValue<Activity>(dialogEvent, "Value");
                         break;
                     }
+
+                case AdaptiveEvents.SetTestOptions:
+                    {
+                        // We received an SetTestOptions event, set value into conversation.testOptions
+                        actionContext.State.SetValue(TestOptions.Kind, dialogEvent.Value);
+                        break;
+                    }
```
Comment (chrimc62) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/AdaptiveDialog.cs:392]: Is this not possible to do with middle ware?  Ideally, we would not change anything in Microsoft.Bot.Builder.Dialogs.Adaptive and the middleware would be part of Adaptive.Testing.  Events support custom payloads an this would be custom.

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/AdaptiveEvents.cs [hunk 8]
```diff
@@ -39,5 +39,10 @@ public class AdaptiveEvents : DialogEvents
         /// Raised when an entity should be assigned to a property.
         /// </summary>
         public const string AssignEntity = "assignEntity";
+
+        /// <summary>
+        /// Set test options.
+        /// </summary>
+        public const string SetTestOptions = "setTestOptions";
```
Comment (chrimc62) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/AdaptiveEvents.cs:46]: Don't add this.  You can do it as a custom event in Adaptive.Testing.

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/TestOptions.cs [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+using System.Collections.Generic;
+using System.Text;
+using Newtonsoft.Json;
+
+namespace Microsoft.Bot.Builder.Dialogs.Adaptive
+{
+    /// <summary>
+    /// Test options.
+    /// </summary>
+    public class TestOptions
```
Comment (chrimc62) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/TestOptions.cs:11]: >TestOptions [](start = 17, length = 11)

Do you actually need this class?  It is not really a $kind in emitEvent.  For the custom event the payload is an arbitrary object, right?  And you can take that arbitrary object and put it into memory without a strongly typed payload.  The nice thing about this is that people can put in any values they like in the test options structure.

File: tests/Microsoft.Bot.Builder.TestBot.Json/Samples/RandomGenerator/RandomGenerator.main.dialog [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    "triggers": [
+        {
+            "$kind": "Microsoft.OnBeginDialog",
+          "actions": [
+            {
+              "$kind": "Microsoft.EmitEvent",
+              "eventName": "setTestOptions",
+              "EventValue": {
+                "randomSeed": 100
+              }
```
Comment (chrimc62) [inline tests/Microsoft.Bot.Builder.TestBot.Json/Samples/RandomGenerator/RandomGenerator.main.dialog:13]: This can be an arbitrary object put into conversation.testOptions.  You don't need to strongly type this.

**Comment by chrimc62** [comment]
Good changes for the RandomValue except for min/max adjustment.
Ideally we should not be changing adaptive at all.  The event is a custom one with an arbitrary payload.  The only place the event should be consumed is in middle ware that is a part of adaptive.testing.  The middle ware should detect the event and put the eventValue into conversation.testOptions.

---
In reply to: [672109749](https://github.com/microsoft/botbuilder-dotnet/pull/4429#issuecomment-672109749) [](ancestors = 672109749)

**Comment by Danieladu** [comment]
> Good changes for the RandomValue except for min/max adjustment.
> Ideally we should not be changing adaptive at all. The event is a custom one with an arbitrary payload. The only place the event should be consumed is in middle ware that is a part of adaptive.testing. The middle ware should detect the event and put the eventValue into conversation.testOptions.
> 
> In reply to: [672109749](https://github.com/microsoft/botbuilder-dotnet/pull/4429#issuecomment-672109749) [](ancestors = 672109749)

Currently, the only thing that is blocked on my side is how to catch the "emitEvent" event in middleware.
I have made some investigations about the "emitEvent" action, and does not find any real `event` was emitted, so, in this way, I can not add a subscriber in a middleware. May you provide some suggestions about this one? Thanks very much.

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/TestOptions.cs [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+    {
+        /// <summary>
+        /// Kind.
+        /// </summary>
+        public const string Kind = "conversation.testOptions";
```
Comment (tomlm) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/TestOptions.cs:16]: remove

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/TestOptions.cs [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+
+        /// <summary>
+        /// Gets or sets random value.
+        /// </summary>
+        /// <value>
+        /// Random value.
+        /// </value>
+        [JsonProperty(PropertyName = "randomValue")]
+        public int RandomValue { get; set; }
+    }
```
Comment (tomlm) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/TestOptions.cs:35]: > } [](start = 3, length = 2)

For C# you want to capture overflow properties

[JsonExtendedData]
public IDictionary<string, object> Properties{get;set;}

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/AdaptiveDialog.cs (hunk 7 above)
Comment (tomlm) [reply libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/AdaptiveDialog.cs:392]: Don't do this here...do it in middleware

---
In reply to: [469416323](https://github.com/microsoft/botbuilder-dotnet/pull/4429#discussion_r469416323) [](ancestors = 469416323)

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Generators/TemplateEngineLanguageGenerator.cs [hunk 13]
```diff
@@ -89,7 +89,17 @@ public TemplateEngineLanguageGenerator(string filePath, Dictionary<string, IList
         /// <returns>generated text.</returns>
         public override Task<object> GenerateAsync(DialogContext dialogContext, string template, object data, CancellationToken cancellationToken = default)
         {
-            var lgOpt = new EvaluationOptions() { Locale = dialogContext.GetLocale() };
+            var lgOpt = new EvaluationOptions()
+            {
+                Locale = dialogContext.GetLocale(),
+            };
+
+            var testOptions = dialogContext.State.GetValue<TestOptions>(TestOptions.Kind);
```
Comment (tomlm) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Generators/TemplateEngineLanguageGenerator.cs:97]: >TestOptions.Kind) [](start = 72, length = 17)

"Conversation.TestOptions"

File: tests/Microsoft.Bot.Builder.TestBot.Json/Samples/RandomGenerator/RandomGenerator.main.dialog [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+    "$schema": "../../testbot.schema",
+    "$kind": "Microsoft.AdaptiveDialog",
+    "triggers": [
+        {
+            "$kind": "Microsoft.OnBeginDialog",
+          "actions": [
+            {
+              "$kind": "Microsoft.EmitEvent",
+              "eventName": "setTestOptions",
+              "EventValue": {
```
Comment (tomlm) [inline tests/Microsoft.Bot.Builder.TestBot.Json/Samples/RandomGenerator/RandomGenerator.main.dialog:11]: >EventValue [](start = 15, length = 10)

lower case eventValue

File: tests/Microsoft.Bot.Builder.TestBot.Json/Samples/RandomGenerator/RandomGenerator.main.dialog (hunk 10 above)
Comment (tomlm) [inline tests/Microsoft.Bot.Builder.TestBot.Json/Samples/RandomGenerator/RandomGenerator.main.dialog:13]: NO, this is wrong direction and wrong eventing.

The testScript needs to send a EventActivity to the BOT, not the BOT=> emitting a dialog event to the stack.

File: libraries/AdaptiveExpressions/BuiltinFunctions/Rand.cs [hunk 15]
```diff
@@ -43,9 +42,28 @@ private static EvaluateExpressionDelegate Evaluator()
...
+                                && randomValue.IsInteger())
+                                {
+                                    var randomValueNum = Convert.ToInt32(randomValue, CultureInfo.InvariantCulture);
+                                    value = min + (randomValueNum % (max - min));
+                                }
+                                else
                                 {
-                                    value = Randomizer.Next(min, max);
+                                    var random = new Random();
+                                    if (options.Properties.TryGetValue("randomSeed", out var randomSeed))
```
Comment (chrimc62) [inline libraries/AdaptiveExpressions/BuiltinFunctions/Rand.cs:54]: Creating a random is relatively expensive and the purpose of a seed is to start a whole sequence of random values.  You also would not need a lock if you create it on the fly.  What you should do: 
1) Have a single static random.
2) If there is no seed or it is a new seed, create a new one.
3) When you use it, do a lock.

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/SetTestOptionsMiddleware.cs [hunk 16]
```diff
@@ -0,0 +1,43 @@
+﻿using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Bot.Schema;
+
+namespace Microsoft.Bot.Builder.Dialogs.Adaptive
+{
+    /// <summary>
+    /// Middleware that catch "SetTestOptions" event and save into "Conversation.TestOptions".
+    /// </summary>
+    public class SetTestOptionsMiddleware : IMiddleware
```
Comment (chrimc62) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/SetTestOptionsMiddleware.cs:10]: >SetTestOptionsMiddleware [](start = 17, length = 24)

This does not belong in Microsoft.Bot.Builder.Dialogs.Adaptive project.  It should be in Microsoft.Bot.Builder.Dialogs.Adaptive.Testing.

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Generators/TemplateEngineLanguageGenerator.cs [hunk 17]
```diff
@@ -89,7 +89,16 @@ public TemplateEngineLanguageGenerator(string filePath, Dictionary<string, IList
...
+            var lgOpt = new EvaluationOptions()
+            {
+                Locale = dialogContext.GetLocale(),
+            };
+
+            var testOptions = dialogContext.State.GetValue<Dictionary<string, object>>("Conversation.TestOptions");
+            if (testOptions != null)
+            {
+                lgOpt.Properties = testOptions;
+            }
```
Comment (chrimc62) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Generators/TemplateEngineLanguageGenerator.cs:101]: I'm not opposed to this, but I suspect that you do not need to copy this into lgOptions. Every place has access to Conversation.TestOptions, right?  If someone wanted to dynamically set this that is where they would do a setProperty.

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive.Testing/TestActions/UserEvent.cs [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+using Microsoft.Bot.Schema;
+using Newtonsoft.Json;
+
+namespace Microsoft.Bot.Builder.Dialogs.Adaptive.Testing.TestActions
+{
+    /// <summary>
+    /// Action to script sending event to the bot.
+    /// </summary>
+    [DebuggerDisplay("UserEvent:{Name}")]
+    public class UserEvent : TestAction
```
Comment (chrimc62) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive.Testing/TestActions/UserEvent.cs:21]: >UserEvent [](start = 17, length = 9)

I suspect a better name would be CustomEvent--it is not actually tied to the user at all.

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/Extensions.cs [hunk 19]
```diff
@@ -147,6 +153,51 @@ public static string GetPrefixErrorMessage(this LGTemplateParser.NormalTemplateS
...
+            }
+
+            if (seed != null)
+            {
+                _random = new Random(seed.Value);
+            }
+            else
+            {
+                if (properties.TryGetValue("randomSeed", out var randomSeed))
+                {
```
Comment (chrimc62) [inline libraries/Microsoft.Bot.Builder.LanguageGeneration/Extensions.cs:182]: Isn't this going to create a new random every time?

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/Extensions.cs [hunk 20]
```diff
@@ -147,6 +153,51 @@ public static string GetPrefixErrorMessage(this LGTemplateParser.NormalTemplateS
...
+        /// If value is not null, the mock random value result would be: min + (value % (max - min)).
+        /// Else if seed is not null, the seed of the random would be fixed.
+        /// </summary>
+        /// <param name="properties">Properties.</param>
+        /// <param name="min">The inclusive lower bound of the random number returned.</param>
+        /// <param name="max">The exclusive upper bound of the random number returned. max must be greater than or equal to min.</param>
+        /// <param name="seed">user seed.</param>
+        /// <returns>Random seed and value.</returns>
+        public static int GeneratorMockRandom(this IDictionary<string, object> properties, int min, int max, int? seed = null)
+        {
```
Comment (chrimc62) [inline libraries/Microsoft.Bot.Builder.LanguageGeneration/Extensions.cs:167]: This should be in Dialog.Adaptive and not in Language Generation.  The reason is dependencies--we don't want selectors depending on LG.

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Selectors/RandomSelector.cs [hunk 21]
```diff
@@ -75,9 +77,18 @@ public override Task<IReadOnlyList<OnCondition>> SelectAsync(ActionContext conte
             if (candidates.Count > 0)
             {
                 int selection;
-                lock (_objectLock)
+                var (value, error) = Expression.Parse("Conversation.TestOptions").TryEvaluate<IDictionary<string, object>>(context.State);
+                if (error == null && value != null)
```
Comment (chrimc62) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive/Selectors/RandomSelector.cs:81]: I don't think test options is an expression.  Conversation.testOptions is just plain old memory like everything else.  Someone could have an expression in SetProperty to compute this, but the memory itself would not be an expression.

File: libraries/Microsoft.Bot.Builder.LanguageGeneration/EvaluationOptions.cs [hunk 22]
```diff
@@ -116,6 +119,16 @@ public EvaluationOptions(IList<string> optionsList)
...
+        /// <summary>
+        /// Gets or sets the additional properties.
+        /// </summary>
+        /// <value>
+        /// Additional properties.
+        /// </value>
+#pragma warning disable CA2227 // Collection properties should be read only
+        public IDictionary<string, object> Properties { get; set; } = new Dictionary<string, object>();
+#pragma warning restore CA2227 // Collection properties should be read only
+
```
Comment (chrimc62) [inline libraries/Microsoft.Bot.Builder.LanguageGeneration/EvaluationOptions.cs:131]: What I don't like about this for the random stuff is I cannot easily change this through doing an action.  If it is just plain memory I can do a SetProperty action.

File: tests/tests.schema [hunk 23]
```diff
@@ -9180,6 +9183,9 @@
         },
         {
           "$ref": "#/definitions/Microsoft.Test.UserSays"
+        },
+                {
+          "$ref": "#/definitions/Microsoft.Test.UserEvent"
         },
         {
```
Comment (chrimc62) [inline tests/tests.schema:9190]: I suspect you did this by hand.  You should not.  What you should do is install the latest bf cli and run the script which creates this file.
